File: java/src/main/java/vsdk/toolkit/io/mgf/MgfEntityHandlerAdapter.java
package vsdk.toolkit.io.mgf;

import vsdk.toolkit.io.context.EntityDispatchContext;
import vsdk.toolkit.io.context.HandlerRoleContext;
import vsdk.toolkit.io.context.ParseContext;
import vsdk.toolkit.io.context.ParseErrorContext;
import vsdk.toolkit.io.context.ParseRuntimeContext;

public final class MgfEntityHandlerAdapter extends EntityDispatchContext {
    @FunctionalInterface
    public interface HandlerFunction {
        int handle(int argc, String[] argv, ParseRuntimeContext context);
    }

    private final HandlerRoleContext handlerType;
    private final HandlerFunction handlerFunction;

    public MgfEntityHandlerAdapter(HandlerRoleContext handlerType, HandlerFunction handlerFunction) {
        this.handlerType = handlerType;
        this.handlerFunction = handlerFunction;
    }

    @Override
    public int handle(int argc, String[] argv, ParseContext context) {
        if (handlerFunction == null) {
            return ParseErrorContext.MGF_OK;
        }
        return handlerFunction.handle(argc, argv, (ParseRuntimeContext)context);
    }

    @Override
    public HandlerRoleContext type() {
        return handlerType;
    }
}

File: typescript/src/main/vsdk/toolkit/io/mgf/MgfEntityHandlerAdapter.ts
import { EntityDispatchContext } from "../context/EntityDispatchContext";
import { HandlerRoleContext } from "../context/HandlerRoleContext";
import { ParseContext } from "../context/ParseContext";
import { ParseErrorContext } from "../context/ParseErrorContext";
import { ParseRuntimeContext } from "../context/ParseRuntimeContext";

export type HandlerFunction = (argc: number, argv: string[], context: ParseRuntimeContext) => number;

export class MgfEntityHandlerAdapter extends EntityDispatchContext {
  private readonly handlerType: HandlerRoleContext;
  private readonly handlerFunction: HandlerFunction | null;

  public constructor(handlerType: HandlerRoleContext, handlerFunction: HandlerFunction | null) {
    super();
    this.handlerType = handlerType;
    this.handlerFunction = handlerFunction;
  }

  public override handle(argc: number, argv: string[], context: ParseContext): number {
    if (this.handlerFunction === null) {
      return ParseErrorContext.MGF_OK;
    }
    return this.handlerFunction(argc, argv, context as ParseRuntimeContext);
  }

  public override type(): HandlerRoleContext {
    return this.handlerType;
  }
}
